TauOSD support

peabody124 opened the pull request:
Provides support TauOSD.

Commit:
ENH: make rssi and home dist label configurable

Conflicts:
	flight/Modules/OnScreenDisplay/onscreendisplay.c
	ground/gcs/src/plugins/config/configosdwidget.cpp
	ground/gcs/src/plugins/config/osdpage.ui

Commit:
OSD: show tablet mode and location

Conflicts:
	flight/Modules/OnScreenDisplay/onscreendisplay.c

Commit:
OSD: fix lla to NED conversion, show tablet as T

peabody124 committed:
TauOSD: initial import

peabody124 committed:
OSD: CAN module for OSD and FC

This module will pass information across the bus.

peabody124 committed:
Sparky2: send OSD messages on CAN

peabody124 committed:
TauOSD: receive CAN messages

peabody124 committed:
OSD CAN Module: send more messages

 - Yaw
 - FlightMode
 - FlightBatteryState

peabody124 committed:
TauOSD: workaround for flash saving errors

peabody124 committed:
FC OSD CAN: convert to thread instead of callback

This was consuming too much time on the eventdispatcher
and resulting in other issues.

peabody124 committed:
OSD CAN: unpack more messages

peabody124 committed:
TauOSD: cleanup and removal of old code

peabody124 committed:
OSD: port upstream changes from Brain

peabody124 committed:
OSD Fonts: define more cleanly

peabody124 committed:
OSD: compass rose and vertical scale font to 1

These should be using a different font but for some
reason it fails when not using font #1. This makes that
a constant for cleaner coding and debugging.

peabody124 committed:
OSD: more work on fonts.

peabody124 committed:
TauOSD: use printf2 to get sprintf

peabody124 committed:
TauOSD: blank OSD on powerup

peabody124 committed:
TauOSD: fix BL

ebernhardson committed:
OSD: Make ArtificialHorizon max pitch configurable

Adds a new parameter ArtificialHorizonMaxPitch to the osd page
configuration and utilizes it in the OnScreenDisplay flight module.
The default value of 30 works for slower flights, but various quads
pitch further than 30 degrees forward resulting in an artificial
horizon that is off-screen.  By making it configurable pilots can
adjust it to match their flight profile, or even set different pages
with different maximum pitches depending on how they are flying at
the moment.

peabody124 committed:
TauOSD: add video select. hardcoded to 0

peabody124 committed:
TauOSD: add hw configuration object

peabody124 committed:
OsdCan: enable battery module when message received

The OSD code uses which modules are enabled as a flag for which
widgets to possibly display. If a battery message is receievd
then we should show this information.

peabody124 committed:
TauLabs Board: add TauOsd to GCS

peabody124 committed:
TauOSD: 256k is plenty for flash

peabody124 committed:
FC OSD: send GPS information when module running

peabody124 committed:
OSD: receive GPS information

peabody124 committed:
OSD: respond to updates to ModuleSettings

This is useful because when running as a standalone OSD the
modules are enabled when messages arrived to make things
"just work".

In addition, getting the sttings at the beginning of Initialize
is very unsafe. This depends on the order modules are initialized
that the object even has been initialized. Never fetch objects
in the Initialize methods.

peabody124 committed:
OSD: send GPS Velocity

peabody124 committed:
OSD: send PositionActual

Commit:
Brain: Add climb rate and G force to OSD

Commit:
OSD support next dedicated mwrate flight mode

Conflicts:
	flight/Modules/OnScreenDisplay/onscreendisplay.c
	flight/Modules/OnScreenDisplay/osd_menu.c

peabody124 committed:
TauOSD: update channel selection for 32ch

peabody124 committed:
TauOSD: can afford to redraw every frame

peabody124 committed:
OSD: no more VelocityControl mode

peabody124 committed:
TauOSD: use open drain outputs for ch selection

peabody124 committed:
TauOSD CAN: send ManualControl RSSI

This is compatible with how Brain sources the data.

Commit:
PIOS Video: fix lockup when no camera attched

Signed-off-by: James Cotton <peabody124@gmail.com>

Commit:
OSD: Fix float operations.

Signed-off-by: James Cotton <peabody124@gmail.com>

Commit:
OSD: work around sprintf bug in pdop display

Signed-off-by: James Cotton <peabody124@gmail.com>

peabody124 committed:
TauOSD: include PIOS_Sensors

dustin committed:
OSD: More detailed alarms.

This is based on the work I did in MSP in TL proper.  It's a slightly
more condensed form to store the names of alarms, but should fail to
compile if new alarms are added and not described here, provides more
detail on system configuration and manual control alarms, and considers
the boot cause to be an alarm for a bit at startup so you can see why it
booted.

Signed-off-by: James Cotton <peabody124@gmail.com>

dustin committed:
OSD: Use new alarm stringing code

This uses common library code to convert alarms to strings
for MWOSD and TauOSD/Brain

Signed-off-by: James Cotton <peabody124@gmail.com>

peabody124 committed:
TauOSD: fix default field value

peabody124 committed:
TauOSD: remove unused UAVOs

peabody124 committed:
TauOSD: printf2 was deleted

droney committed:
flight: printf2 stack alignment change

peabody124 committed:
OsdCAN: process the alarm messages from FC to OSD

mlyle committed:
flight: replace _sbrk implementation with failure

Signed-off-by: James Cotton <peabody124@gmail.com>

kubark42 commented:
@peabody124 Do you mind if I break up some of this into smaller PRs? For instance, I'd like to review the sine table separately, and there's some work which could probably merge in as is such as mgrs.c. I'm happy to have those not have homes for the moment until the TauOSD stuff gets merged.\* And do you think the CAN stuff is separable from the OSD?

[*]It also makes it easier to catch things like including the brain Makefile. :D

peabody124 commented:
I can take a crack at doing this. It might make it easiet to preserve more of the original commits. Did you just want CAN v OSD or more granular?

kubark42 commented:
Whatever is easiest, with a preference toward granularity. It makes it much easier to pull that code into CyPhy's codebase-- or internally revert it if it causes merge conflicts we don't want to deal with immediately.

If there are math libraries that can be reviewed separately I will be able to get them in ASAP. With the sine library, we might want to address https://github.com/TauLabs/TauLabs/issues/477 if it's still an outstanding issue.

peabody124 committed:
OSD: use sinf/cosf instead of lookup

peabody124 commented:
This has been rebased on #2124 and #2125 which should be merged first.

kubark42 commented:
> This has been rebased on #2124 and #2125 which should be merged first.

@peabody124 Thanks a lot, really appreciate it. 

Looks like we can get #2124 and #2125 in with minimal change so I hope to be able to review this soon.

peabody124 committed:
TauOSD/Sparky2: support pios_can_common

peabody124 committed:
TauOSD: no more sin_lookup

mluessi committed:
OSD: fix alarm display

Signed-off-by: James Cotton <peabody124@gmail.com>

peabody124 committed:
OSD: support AcroPlus

peabody124 committed:
OSD: add dR tag in author list

peabody124 committed:
OSD: add blink flag

peabody124 committed:
TauOSD: enable battery monitoring on PC3

To make life easier on users, when TauOSD has the battery
module enabled, it correctly configures the pin and scale.
They can still specify voltage thresholds if desired but
the alarms might play strangely with those sourced from the
main flight controller.

peabody124 committed:
OSD CAN: enable bidirectional sending of battery voltage

Since the OSD can monitor the battery voltage, it is nice to
be able to send that back to the FC. This detects which side
has the module running and sends it to the other side. It
anticipates both sides will not simultaneously have it running.

kubark42 commented:
@peabody124 Could you rebase this onto latest next? Hopefully that will make Jenkins happy.

peabody124 commented:
Will do tonight!

kubark42 commented:
@peabody124 poke on rebasing this.

peabody124 commented:
@kubark42 thanks for poke. I tried this morning but there was a conflict so I didn't want to gloss over it while rushing to work :) Should be quick and will do later.

kubark42 commented:
Awesome. Looking forward to getting this in.

peabody124 commented:
Rebased. Note that by necessity this currently included #2126 and this weekend I'll take a crack at generalizing that. No promises though.

peabody124 committed:
TauOSD: enable FastHeap

kubark42 commented:
Thanks. If #2126 can't be generalized easily, we'll push forward with it as is.

P.S. Jenkins hiccups on `/var/lib/jenkins/workspace/taulabs-pr-linux-amd64/ground/gcs/src/plugins/boards_naze/naze.cpp:71:11: error: enumeration value ‘BOARD_CAPABILITIES_OSD’ not handled in switch [-Werror=switch]`

peabody124 committed:
OSD CAN: add task info for OSD CAN comms

peabody124 committed:
TauOSD: convert to ChiBios

peabody124 committed:
OSD: take home_baro from first sample

Because the comms might take a few seconds to update
the baro, otherwise it averages zeros into the running
average.

peabody124 committed:
TauOSD: do not enable CAN PIOS_COM layer

peabody124 committed:
OSD CAN: remove unused define

peabody124 committed:
MSPUavoBridge: check in stub

This is a module for the OSD that will systematically make
MSP queries to update the display. This is based off the
opposite end - the UAVOMSPBridge which responds to these
queries from a MWOSD.

At some point, factoring the common MSP parsing code into a
library will make a lot of sense.

peabody124 committed:
MSPUavoBridge: more work

peabody124 committed:
TauOSD: enable MSPUavoBridge

peabody124 committed:
PIOS HAL: compile when no receiver enabled

peabody124 committed:
TauOSD: support PIOS_HAL

peabody124 committed:
TauOSD: enable msplib

peabody124 committed:
TauOSD: fix USART configuration

peabody124 committed:
MSPUavoBridge: more work to use MSP LIB

peabody124 committed:
MSPUavoBridge: handle more packets

peabody124 committed:
MSPUavoBridge: use cleaner scheduling system

This makes it easier to set up the timings and will be
adjustable based on baudrate.

peabody124 committed:
MSPUAVOBridge: update callback definition

kubark42 commented:
Since #2126 is merged, want to rebase this for final review?

peabody124 committed:
GCS/Boards: indicate that they do not support an OSD

peabody124 commented:
@kubark42 done. pushed build fix (it was just a warning on OS X)

kubark42 commented:
Still one last warning: 

```
ground/gcs/src/plugins/boards_taulabs/sparky.cpp:80:11: error: enumeration value ‘BOARD_CAPABILITIES_OSD’ not handled in switch [-Werror=switch]
13:44:11      switch(capability) {
```

peabody124 committed:
TauOSD: enable DAC BEEP

peabody124 commented:
@kubark42 fixed

kubark42 commented on flight/targets/brain/fw/Makefile:
```diff
@@ -0,0 +1,347 @@
+###############################################################################
+# @file       Makefile
```
This file doesn't belong here, right?

kubark42 commented on flight/Modules/OnScreenDisplay/inc/osd_utils.h:
```diff
@@ -0,0 +1,177 @@
+#include "misc_math.h"
+
+// Size of an array (num items.)
+#define SIZEOF_ARRAY(x) (sizeof(x) / sizeof((x)[0]))
```
There's a `#define NELEMENTS(x) (sizeof(x) / sizeof(*(x)))` defined in `pios.h`. (It's actually defined in many, many places and we might want to consider unifying that into our `shared/api` directory.)

kubark42 commented on flight/Modules/OnScreenDisplay/inc/osd_utils.h:
```diff
@@ -0,0 +1,177 @@
+#define CLIP_COORD_Y(y)              { y = y < GRAPHICS_TOP ? GRAPHICS_TOP : y > GRAPHICS_BOTTOM ? GRAPHICS_BOTTOM : y; }
+
+// Macro to swap two variables using XOR swap.
+#define SWAP(a, b)                   { a ^= b; b ^= a; a ^= b; }
```
That's a pretty cool macro.

kubark42 commented on flight/Modules/OnScreenDisplay/inc/osd_utils.h:
```diff
@@ -0,0 +1,177 @@
+
+// Text dimension structures.
+struct FontDimensions {
+	int width, height;
```
Not that it particularly bothers me, but this would have been better as the more pedantic:

```
struct FontDimensions {
    int width;
    int height;
}
```

kubark42 commented on flight/Modules/OnScreenDisplay/onscreendisplay.c:
```diff
@@ -0,0 +1,1448 @@
+// Unit conversion constants
+#define MS_TO_KMH 3.6f
+#define MS_TO_MPH 2.23694f
+#define M_TO_FEET 3.28084f
```
Put these into `physical_constants.h`. (Except `MS_TO_KMH` which is already defined there.)

kubark42 commented on flight/Modules/OnScreenDisplay/onscreendisplay.c:
```diff
@@ -0,0 +1,1448 @@
+
+// static assert
+#define ASSERT_CONCAT_(a, b) a##b
+#define ASSERT_CONCAT(a, b) ASSERT_CONCAT_(a, b)
```
What's the point of the double (and redundant?) MACRO for concatenation? Could that be accomplished in just one MACRO?

kubark42 commented on flight/Modules/OnScreenDisplay/onscreendisplay.c:
```diff
@@ -0,0 +1,1448 @@
+#include "taskinfo.h"
+#include "velocityactual.h"
+#include "waypoint.h"
+#include "waypointactive.h"
```
Thanks for alphabetizing!

kubark42 commented on flight/Modules/OnScreenDisplay/onscreendisplay.c:
```diff
@@ -0,0 +1,1448 @@
+ * @param       majtick_len     major tick length
+ * @param       flags           special flags (see hud.h.)
+ */
+#define COMPASS_SMALL_NUMBER
```
Could be cool to have a comment describing what these `#defines` do.

kubark42 commented on flight/Modules/OnScreenDisplay/onscreendisplay.c:
```diff
@@ -0,0 +1,1448 @@
+	const int FONT_ID = 0;
+	v %= 360; // wrap, just in case.
+	struct FontEntry font_info;
+	int majtick_start = 0, majtick_end = 0, mintick_start = 0, mintick_end = 0, textoffset = 0;
```
Hmmm, is this copied from somewhere else? This doesn't look like our normal style. I can't be for sure, but didn't @solidgoldbomb convince us that stylistically we should have every variable be defined on its own line?

peabody124 replied:
yeah this part of the code was heavily overhauled by @BrainFPV from the older code
we had. i might do some modification to it but may leave the major stuff for later to try
and keep it more compatible for back porting changes.

kubark42 commented on flight/Modules/OnScreenDisplay/onscreendisplay.c:
```diff
@@ -0,0 +1,1448 @@
+	uint8_t last_page = -1;
+
+	OnScreenDisplaySettingsGet(&osd_settings);
+	home_baro_altitude = 0.;
```
stray `.`

kubark42 commented on flight/Modules/OnScreenDisplay/onscreendisplay.c:
```diff
@@ -0,0 +1,1448 @@
+
+				if (osd_settings.Units == ONSCREENDISPLAYSETTINGS_UNITS_IMPERIAL){
+					convert_distance = M_TO_FEET;
+					convert_distance_divider = 5280.f; // feet in a mile
```
This should be added into `physical_constants.h`. There's already the inverse conversion, `FEET2MILES`, defined.

kubark42 commented on flight/Modules/OnScreenDisplay/onscreendisplay.c:
```diff
@@ -0,0 +1,1448 @@
+	home_baro_altitude = 0.;
+
+	// blank
+	while (PIOS_Thread_Systime() <= BLANK_TIME) {
```
This needs better documentation. What's the value of this? Is it just to keep the OSD from showing anything before the user has time to put on the googles?

kubark42 commented on flight/PiOS/STM32F4xx/pios_video.c:
```diff
@@ -0,0 +1,567 @@
+void PIOS_First_Line_ISR(void)
+{
+	if(TIM_GetITStatus(dev_cfg->line_counter, TIM_IT_Update) && (active_line == 0))
+	{
```
`{`

peabody124 committed:
PhysicalConstants: add constants used by OSD

peabody124 committed:
OSD: changes from review

peabody124 committed:
OSD: bump copyright

peabody124 committed:
TauOSD: enable FSKDAC telemetry

peabody124 committed:
TauOSD: more work enabling FSKDAC

peabody124 committed:
OSD CAN: do not run if no port

peabody124 committed:
OSD CAN: pass and receiver vertical message

peabody124 committed:
TauOSD: set LTM to 1200bps when FSK

kubark42 commented:
Is there any work left on this to be done? Would you like to merge now or wait for https://github.com/TauLabs/TauLabs/pull/2209 and update it to be compatible with that?

peabody124 commented:
@kubark42 ok with me

peabody124 committed:
TauOSD/FC: fix crash for battery listening

peabody124 committed:
OSD Menu: fix for arming settings

peabody124 committed:
TauOSD: fix BL compilation
